src/storage/src/source/generator/upsert.rs

```diff
@@ -0,0 +1,556 @@
+            config.source_resume_uppers[&config.id]
+                .iter()
+                .map(MzOffset::decode_row),
+        );
+
+        let Some(resume_offset) = resume_upper.into_option() else {
+            return;
+        };
+
+        start_signal.await;
+
+        let snapshotting = resume_offset.offset == 0;
+
+        let mut local_partitions: BTreeMap<_, _> = (0..upsert.partitions)
+            .into_iter()
+            .filter_map(|p| {
+                config.responsible_for(p).then(|| {
+                    (
+                        p,
+                        SnapshotProducer::new(
+                            p,
+                            upsert.partitions,
+                            upsert.keys,
+                            upsert.snapshot_rounds,
+                            upsert.batch_size,
+                            upsert.seed,
+                        ),
+                    )
+                })
+            })
+            .collect();
+
+        let stats_worker = config.responsible_for(0);
+
+        if local_partitions.is_empty() {
+            stats_output
+                .give(
+                    &stats_cap,
+                    ProgressStatisticsUpdate::Snapshot {
+                        records_known: 0,
+                        records_staged: 0,
+                    },
+                )
+                .await;
+            return;
+        }
+
+        let local_snapshot_size =
+            (u64::cast_from(local_partitions.len())) * upsert.keys * upsert.snapshot_rounds
+                / upsert.partitions;
+
+        // Re-usable buffers.
+        let mut value_buffer: Vec<u8> = vec![0; usize::cast_from(upsert.value_size)];
+        let mut updates_buffer = Vec::new();
+
+        // snapshotting
+        let mut upper_offset = if snapshotting {
+            if stats_worker {
+                stats_output
+                    .give(
+                        &stats_cap,
+                        ProgressStatisticsUpdate::SteadyState {
+                            offset_known: upsert.snapshot_rounds,
+                            offset_committed: 0,
+                        },
+                    )
+                    .await;
+            };
+
+            // Downgrade to the
+            progress_cap.downgrade(&MzOffset::from(upsert.snapshot_rounds));
+
+            let mut emitted = 0;
+            stats_output
+                .give(
+                    &stats_cap,
+                    ProgressStatisticsUpdate::Snapshot {
+                        records_known: local_snapshot_size,
+                        records_staged: emitted,
+                    },
+                )
+                .await;
+            while local_partitions.values().any(|si| !si.finished()) {
+                for sp in local_partitions.values_mut() {
+                    updates_buffer.clear();
+                    emitted += sp.produce_batch(&mut updates_buffer, &mut value_buffer);
```

Comment:
if we produced 0 we could early continue here, but just a nit

Reply:
I think its probably clearer this way!